update Observability with SharedTelemetryConfig (#376)
upbound/docs: content/spaces/observability.md

## SharedTelemetryConfig

SharedTelemetryConfig is a custom resource that defines the telemetry configuration for a group of control planes. It specifies the exporters and pipelines that the control planes in the group use to send telemetry data to external observability backends.

 The following is an example of a SharedTelemetryConfig resource that sends metrics and traces to New Relic:

```yaml
apiVersion: observability.spaces.upbound.io/v1alpha1
kind: SharedTelemetryConfig
metadata:
  name: newrelic
  namespace: default
spec:
  controlPlaneSelector:
    labelSelectors:
      - matchLabels:
          org: foo
  exporters:
    otlphttp:
      endpoint: https://otlp.nr-data.net
      headers:
        api-key: YOUR_API_KEY
  exportPipeline:
    metrics: [otlphttp]
    traces: [otlphttp]
```

The `controlPlaneSelector` field specifies the control planes that use this configuration.
The `exporters` field specifies the configuration for the exporters. The configuration for each exporter is specific to the exporter and corresponds to its [OpenTelemetry Collector configuration](https://opentelemetry.io/docs/collector/configuration/#exporters).

> The `exportPipeline` field specifies the pipelines that the control planes use to send telemetry data to the exporters. The `metrics` and `traces` fields specify the names of the pipelines that the control planes use to send metrics and traces, respectively. The names of the pipelines correspond to the `exporters` in the OpenTelemetry Collector [service pipeline configuration](https://opentelemetry.io/docs/collector/configuration/#pipelines).

Reviewer: ```suggestion
The `exportPipeline` field specifies the control plane pipelines that send telemetry data to the exporters. The `metrics` and `traces` fields specify the names of the pipelines the control planes use to send metrics and traces, respectively. The names of the pipelines correspond to the `exporters` in the OpenTelemetry Collector [service pipeline configuration](https://opentelemetry.io/docs/collector/configuration/#pipelines).
```

## Usage

In a Space, you can configure per control plane group telemetry settings by creating one or more SharedTelemetryConfig resources.

{{< hint "important" >}}
Only one SharedTelemetryConfig can be applied to a control plane. If multiple SharedTelemetryConfig resources are applied to the same control plane, the one applied first takes precedence. The other SharedTelemetryConfig will consider the control plane telemetry provisioning as failed due to conflict.
{{< /hint >}}

Currently supported exporters are:

> - `datadog` (check [docs](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/exporter/datadogexporter/README.md) for configuration details)

Reviewer: ```suggestion
- `datadog` (review the OpenTelemetry [documentation](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/exporter/datadogexporter/README.md) for configuration details)
```

> - `otelhttp` (used by New Relic among others, check [docs](https://docs.newrelic.com/docs/more-integrations/open-source-telemetry-integrations/opentelemetry/get-started/opentelemetry-set-up-your-app/) for configuration details)

Reviewer: ```suggestion
- `otelhttp` (used by New Relic among others, review the New Relic [documentation](https://docs.newrelic.com/docs/more-integrations/open-source-telemetry-integrations/opentelemetry/get-started/opentelemetry-set-up-your-app/) for configuration details)
```

The example below shows how to configure a SharedTelemetryConfig resource to send metrics and traces to Datadog:

```yaml
apiVersion: observability.spaces.upbound.io/v1alpha1
kind: SharedTelemetryConfig
metadata:
  name: datadog
  namespace: default
spec:
  controlPlaneSelector:
    labelSelectors:
      - matchLabels:
          org: foo
  exporters:
    datadog:
      api:
        site: ${DATADOG_SITE}
        key: ${DATADOG_API_KEY}
  exportPipeline:
    metrics: [datadog]
    traces: [datadog]
```

### Status

If all goes good, the SharedTelemetryConfig will be created and the OpenTelemetry Collector will be provisioned for the selected controlplane. To see the status:

```bash
 kubectl get stc
NAME       SELECTED   FAILED   PROVISIONED   AGE
datadog    1          0        1             63s
```

- `SELECTED` shows the number of control planes selected by the SharedTelemetryConfig.
- `FAILED` shows the number of control planes that failed to provision the OpenTelemetry Collector.

> - `PROVISIONED` shows the number of control planes that successfully provisioned the OpenTelemetry Collector, and the OpenTelemetry Collector is running.

Reviewer: ```suggestion
- `PROVISIONED` shows the provisioned and running OpenTelemetry Collectors on each control plane.
```

> To get more data, we can check the resource status, in which we can see the names of the control planes that were selected and provisioned:

Reviewer: ```suggestion
To return the names of control planes selected and provisioned, review the resource status:
```

```yaml
...
status:
  selected:
    - ctp
  provisioned:
    - ctp
```

> If a conflict or another issue occurs, the failed control planes status will show the failure conditions:

Reviewer: ```suggestion
If a conflict or another issue occurs, the failed control planes status returns the failure conditions:
```

```bash
k get stc
NAME       SELECTED   FAILED   PROVISIONED   AGE
datadog    1          1        0             63s
```

```yaml
...
status:
  failed:
  - conditions:
    - lastTransitionTime: "2024-04-26T09:32:28Z"
      message: 'control plane dev is already managed by another SharedTelemetryConfig:
        newrelic'
      reason: SelectorConflict
      status: "True"
      type: Failed
    controlPlane: ctp
  selectedControlPlanes:
  - ctp
```

> Upbound marks the control plane as provisioned only if the OpenTelemetry Collector is successfully deployed and running. There could be a delay in the status update if the OpenTelemetry Collector is still being deployed, such as:

Reviewer: ```suggestion
Upbound marks the control plane as provisioned only if the OpenTelemetry Collector is deployed and running. There could be a delay in the status update if the OpenTelemetry Collector is currently deploying:
```

```bash
 k get stc
NAME       SELECTED   FAILED   PROVISIONED   AGE
datadog    1          0        0             63s
```

## Prerequisites

> This feature requires the [OpenTelemetry Operator](https://opentelemetry.io/docs/kubernetes/operator/) to be installed on the Space cluster. Install this now if you haven't already:

Reviewer: ```suggestion
This feature requires the [OpenTelemetry Operator](https://opentelemetry.io/docs/kubernetes/operator/) on the Space cluster. Install this now if you haven't already:
```
